### PR TITLE
Added formspec element 'KeyEventBox' for charbased input

### PR DESCRIPTION
**Pull request by ninnghazad**
This adds a new formspec-type that when focused, grabs keyboard input and immediatly
sends it as keycode. This will enable charbased input for mods. I want to use this
for ingame computers in dronetest-mod. Could also be used for lots of other stuff
like musical instruments, remote controls and everything that doesn't do with linebased
input. It can be unfocused by mouse. It is based on a hidden editbox, so it should work 
with the software-keyboard.
it is used like this:
proxy[pos;size;name;texture;active-texture]

Also i changed the formspec-drawing to have a dirty-variable,
which gets set on change in focus. This enables the proxy element (and possibly others)
to have immediate visual reaction to focus change.

### CODE REVIEW

**Comment by rubenwardy** [comment]
Why did you rebase but then close the problem?

**Comment by ninnghazad** [comment]
Wanted to get it working again, so i can play with it. Then remembered there was this ancient pullrequest, which i considered dead by now. Can open again, but would test rebased version some more first.

**Comment by obneq** [comment]
please do!

**Comment by ninnghazad** [comment]
ok. so it works again.
i am however a little unsure about how to go about deactivating the keyboard-input, which would mean unfocusing the element. 
currently you can click another element in the formspec, or double-click to close the formspec to stop typing. 
however i would prefer if i could click the element when it is focused in order to unfocus it, like a toggle-button. couldn't find examples for the in source on quick look though - seems unfocusing without focusing something else is not really used.
ideas in that direction are welcome.

**Comment by ninnghazad** [comment]
close/reopen just to trigger travis rebuild.

**Comment by ninnghazad** [comment]
another rebase and a bump.
what's the status on accepting this?
is there anything else i can do to get this through?

File: src/guiFormSpecMenu.cpp [hunk 1]
```diff
@@ -2892,6 +2999,31 @@ static bool isChild(gui::IGUIElement * tocheck, gui::IGUIElement * parent)
...
+	if (event.EventType == EET_MOUSE_INPUT_EVENT && event.MouseInput.Event == EMIE_LMOUSE_PRESSED_DOWN) {
+		gui::IGUIElement *focused = Environment->getFocus();
+		if(focused && isMyChild(focused)) {
+			if(getTypeByID(focused->getID()) == f_KeyEventBox) {
+				std::string empty;
+				this->setFocus(empty);
+				//return true;
+			}
+		}
+	}*/
```
Comment (Ekdohibs) [inline src/guiFormSpecMenu.cpp]: Why do you add commented out code?

File: src/guiFormSpecMenu.cpp [hunk 2]
```diff
@@ -2892,6 +2999,31 @@ static bool isChild(gui::IGUIElement * tocheck, gui::IGUIElement * parent)
 
 bool GUIFormSpecMenu::preprocessEvent(const SEvent& event)
 {
+	if(event.EventType==EET_KEY_INPUT_EVENT && (event.KeyInput.PressedDown)) {
```
Comment (Ekdohibs) [inline src/guiFormSpecMenu.cpp]: Spaces around operator

File: src/guiFormSpecMenu.cpp [hunk 3]
```diff
@@ -2372,13 +2473,19 @@ void GUIFormSpecMenu::drawSelectedItem()
...
 			m_formspec_string = newform;
-			regenerateGui(m_screensize_old);
+			regenerate = true;
 		}
+		}
+	if(m_dirty) {
+		regenerate = true;
+		m_dirty = false;
 	}
+	if(regenerate) regenerateGui(m_screensize_old);
```
Comment (Ekdohibs) [inline src/guiFormSpecMenu.cpp]: Do not put the `if` body on the same line please.

File: src/guiFormSpecMenu.cpp [hunk 4]
```diff
@@ -1887,6 +1982,12 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
 		GUITable *table = m_tables[i].second;
 		mydata.table_dyndata[tablename] = table->getDynamicData();
 	}
+	
+	//preserve editboxes
+	for (u32 i=0; i<m_fields.size(); i++) {
+		if(getElementFromId(m_fields[i].fid)->getTypeName() != std::string("editBox")) continue;
```
Comment (Ekdohibs) [inline src/guiFormSpecMenu.cpp]: Spaces after `if` and around operators; also do not put the body on the same line.

**Comment by ninnghazad** [comment]
ping - please review changes from february

File: src/guiFormSpecMenu.cpp [hunk 5]
```diff
@@ -2217,15 +2320,15 @@ void GUIFormSpecMenu::drawList(const ListDrawSpec &s, int phase,
 	video::IVideoDriver* driver = Environment->getVideoDriver();
 
 	Inventory *inv = m_invmgr->getInventory(s.inventoryloc);
-	if(!inv){
+	if (!inv){
```
Comment (SmallJoker) [inline src/guiFormSpecMenu.cpp]: Well, thanks for correcting the code style but either fix the whole line or un-touch this code.
`if (condition) {` Brace on the same, with spaces around the outer paranteses.

**Comment by ninnghazad** [comment]
Applied changes requested by @SmallJoker .

**Comment by nerzhul** [comment]
wow a three years old pr, you are here :) snouds nice

File: src/guiFormSpecMenu.cpp [hunk 6]
```diff
@@ -1811,11 +1901,16 @@ void GUIFormSpecMenu::parseElement(parserData* data, std::string element)
...
 		parseField(data,description,type);
 		return;
 	}
 
+	if (type == "keyeventbox") {
+		parseKeyEventBox(data,description);
+		return;
+	}
+
 	if (type == "label") {
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: else if

File: src/guiFormSpecMenu.cpp [hunk 7]
```diff
@@ -1887,6 +1982,14 @@ void GUIFormSpecMenu::regenerateGui(v2u32 screensize)
 		GUITable *table = m_tables[i].second;
 		mydata.table_dyndata[tablename] = table->getDynamicData();
 	}
+	
+	//preserve editboxes
+	for (u32 i=0; i<m_fields.size(); i++) {
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: missing spaces

File: src/guiFormSpecMenu.cpp [hunk 8]
```diff
@@ -2201,7 +2304,7 @@ GUIFormSpecMenu::ItemSpec GUIFormSpecMenu::getItemAtPos(v2s32 p) const
 			s32 y = (i/s.geom.X) * spacing.Y;
 			v2s32 p0(x,y);
 			core::rect<s32> rect = imgrect + s.pos + p0;
-			if(rect.isPointInside(p))
+			if (rect.isPointInside(p))
 			{
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: place move brace on the right line

File: src/guiFormSpecMenu.cpp [hunk 9]
```diff
@@ -3793,3 +3927,19 @@ std::wstring GUIFormSpecMenu::getLabelByID(s32 id)
...
 }
+
+/**
+ * get ftype of element by element id
+ * @param id of element
+ * @return FormSpecFieldType or f_Unknown
+ */
+FormspecFieldType GUIFormSpecMenu::getTypeByID(s32 id)
+{
+	for(std::vector<FieldSpec>::iterator iter =  m_fields.begin();
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: missing spaces

Comment (ninnghazad) [inline src/guiFormSpecMenu.cpp]: The other spots i got, but this one i don't quite understand. What's wrong with the spaces here? Basically made it look like the two similar functions above it.

**Comment by ninnghazad** [comment]
@Ekdohibs @nerzhul please accept/deny changes or request more changes. (commented on the last of the 4 requests by @nerzhul)

File: src/guiFormSpecMenu.cpp [hunk 10]
```diff
@@ -3785,11 +3864,27 @@ std::string GUIFormSpecMenu::getNameByID(s32 id)
  */
 std::wstring GUIFormSpecMenu::getLabelByID(s32 id)
 {
-	for(std::vector<FieldSpec>::iterator iter =  m_fields.begin();
+	for(std::vector<FieldSpec>::iterator iter = m_fields.begin();
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: you need to add a space after for keyword. You don't need to fix the whole file

**Comment by ninnghazad** [comment]
Ok, changed those spaces and fixed a few more braces i found along the way.

File: src/guiFormSpecMenu.cpp [hunk 11]
```diff
@@ -3641,6 +3692,14 @@ bool GUIFormSpecMenu::OnEvent(const SEvent& event)
 				return true;
 			}
 		}
+		if (event.GUIEvent.EventType==gui::EGET_ELEMENT_FOCUSED && isVisible()) {
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: missing spaces around == here too

**Comment by ninnghazad** [comment]
K, fixed those too.

File: src/guiFormSpecMenu.cpp [hunk 12]
```diff
@@ -3785,11 +3844,27 @@ std::string GUIFormSpecMenu::getNameByID(s32 id)
...
 }
+
+/**
+ * get ftype of element by element id
+ * @param id of element
+ * @return FormSpecFieldType or f_Unknown
+ */
+FormspecFieldType GUIFormSpecMenu::getTypeByID(s32 id)
+{
+	for (std::vector<FieldSpec>::iterator iter = m_fields.begin();
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: const_iterator please

File: src/guiFormSpecMenu.cpp [hunk 13]
```diff
@@ -3785,11 +3844,27 @@ std::string GUIFormSpecMenu::getNameByID(s32 id)
  */
 std::wstring GUIFormSpecMenu::getLabelByID(s32 id)
 {
-	for(std::vector<FieldSpec>::iterator iter =  m_fields.begin();
+	for (std::vector<FieldSpec>::iterator iter = m_fields.begin();
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: same here

**Comment by ninnghazad** [comment]
Ok, changed to constant iterators as requested.

**Comment by ShadowNinja** [comment]
We discussed this in the developer meeting and thought it would be better implemented client-side, as two network round trips and a server step and going to make this rather laggy.

**Comment by ninnghazad** [comment]
thanks for taking the time to discuss this pr. 
the fact that this is implemented server-side is however not accidential. rather that i need it to be.  
let me try to explain how this came to be.
back in the day I really enjoyed a minecraft-mod called computercraft. especially a programmable block called turtle, that could move, mine, place other blocks... 
when i  came across minetest i thought it'd be sweet to have something like computercraft. so dronetest was started. it has computers and drones (turtles) that are controlled by userscripts running on the server. the computers provide a basic shell and means to write and run scripts. computers and drones are able to interact with the world. userinput (with this pr)  is displayed on the computer-blocks screen, well, like on an old terminal computer, for every player to see that is looking at it. one could write a script to move the drone using the keyboard. or to use userinput to control connected devices from other mods (digilines and mesecons I think, has been a while). lots of possibilities. to have that "real" computer feeling and be able to use the keyboard, i had to write this pr in order to be able to process arbitrary user keypresses on the server. linebased input and textfieldbased upload were considered and did not come close to the feel of typing on a terminal. ie there is a minimal texteditor, a bit like nano, something like that isn't possible without individual keypresses.
so while I understand the concerns, there isn't really any doing it client-side, because doing it server-side is the purpose of this. it's just what it is supposed to do.
however I can understand if this kind of feature isn't wanted, seems not many people aside from me have had a need for this the last years. 
just wanted to make clear that this would be server-sided by intend.

**Comment by electrodude** [comment]
Instead of remaining fixed on doing things the ComputerCraft way, why don't you try to find some other way that will fit into Minetest better?  For ages, people have been talking about adding client-side lua.  If you tried to implement client-side lua yourself, or if you found some old attempt and finished it, or if I find the time to do it eventually, it would be a great way to implement a computer mod.

Once client-side lua is in place, you could make two types of computer: client-side ones (think smartphones) with proper UIs (like what you're proposing here, except all client-side) that run on the client of the player who has it in his inventory, and server-side ones (think server computers) with no UIs (right-click just shows a formspec with basic settings like digiline channels) that are accessed primarily from client-side computers by some sort of remote login scheme.  This also has other interesting advantages, such as that players would be able to set the debug.sethook timeout as high as they like on their own clients, or even disable it, and not have to face the time limit set by the server - it can't do anything bad other than hang the client-side lua thread on the player's own computer.

You would have to work out what happens to client-side computers that aren't in anyone's inventory.  If a client-side computer changes hands, does it get rebooted, or instead does its state get serialized, shipped to the other player, and then deserialized (libraries such as [pluto](https://github.com/hoelzro/pluto) exist to do this rather scary task)?

**Comment by rubenwardy** [comment]
We already have client side Lua, although you can't transfer mods from a server to a client yet

**Comment by paramat** [comment]
IRC:
19:17 rubenwardy  I feel like that's quite easy to abuse
19:17 rubenwardy  ie: sending formspec updates on key presses
19:17 nerzhul     it's an interesting thing for formspec based games ?
19:18 rubenwardy  well: latency
19:18 rubenwardy  this is basically what CSM was designed for, IMO - advanced GUIs and input, with low latency
19:18 paramat     https://github.com/minetest/minetest/pull/1737#issuecomment-296393469 already discussed
19:19 nerzhul     rubenwardy, ofc
19:19 paramat     personally it seems quite specialised and due to issues, i'm not keen
19:20 nerzhul     just pass, next issue ?
19:22 paramat     well, it was discussed last week and decided against, there are roughly 3 -1s, close or 'improvement needed'?
19:23 paramat     well we can close in a while if not changed i guess

**Comment by ninnghazad** [comment]
eh, too bad. won't change, so closing this.

**Pull request closed without merging**